### PR TITLE
Start fixing tests

### PR DESCRIPTION
**Pull request by tsibley**
Tests seem to have been failing for a while on `develop`. :worried:  This PR starts to fix them. Details are in the commit messages, but so far this:

1. Addresses a few missing `tests/golden/` files.
2. Adds a test runner for running tests individually so that the extent of failures can be seen more easily.

My motivation for starting this work is that I wanted to fix a tricky bug, but I couldn't use the test suite to see if the fix unexpectedly broke things. I'll be submitting my proposed fix in a separate PR.

### CODE REVIEW

**Review by anjakefala — CHANGES_REQUESTED** [review]
Hi @tsibley!

I really appreciate that you underwent this work, but unfortunately, I am going to have to say no to most of this PR. 

`dev/test.sh` contains already existing functionality to run the tests individually. 

```
dev/test.sh name-of-test
```

Example:

```
dev/test.sh errors
```

will run `tests/errors.vd`.

Also, I am sorry that it made it difficult for you to check your fixes, but those goldens had not been added for a reason. I had a note to review those tests with @saulpw before adding them. (errors.tsv especially looked really funny.) We have not had time to look at them, yet. =(

I would rather not have those tests passing right now, I think they might be falslely passing.

Can you please check if the `dev/test.sh` does what you added `dev/tests-individually.sh` for? If it does, can you remove `dev/tests-individually.sh` from the PR. I see that you do have some added niceties in your test-script. Maybe consider adding those to `dev/test.sh`?

Please remove the commit with the added goldens regardless.

In general, updating the goldens for a test is welcome, but please add that you check that the functionality they are looking at works before you do so!

**Comment by tsibley** [comment]
Hi @anjakefala, thanks for the quick review! I really appreciate your feedback. :-)

> Also, I am sorry that it made it difficult for you to check your fixes, but those goldens had not been added for a reason. I had a note to review those tests with @saulpw before adding them. (errors.tsv especially looked really funny.) … I would rather not have those tests passing right now, I think they might be falslely passing.

Nod. That makes sense. I wondered about the goldens not being correct (esp. since they aren't valid TSVs), but reasoned that before 56723a9 those tests were considered "passing" because their output wasn't compared to anything. So I took their output pre-56723a9 and called it the golden. I'll remove those files.

> `dev/test.sh` contains already existing functionality to run the tests individually.

Yep, I know. :-) You'll see that `dev/test-individually.sh` calls `dev/test.sh` for each file.

> Can you please check if the `dev/test.sh` does what you added `dev/tests-individually.sh` for? If it does, can you remove `dev/tests-individually.sh` from the PR.

It doesn't quite do what I need. The reason I wrote `dev/tests-individually.sh`—initially just for my own use, I wasn't planning to commit it—is that it runs each test regardless of if some of them fail. `dev/test.sh` aborts on the first failure, so you don't see later failures until you fix the first one. I first wrote a quick loop on the command-line over `./dev/test.sh` while I was bug hunting and that loop then became `dev/tests-individually.sh`.

> I see that you do have some added niceties in your test-script. Maybe consider adding those to `dev/test.sh`?

I'm happy to work on improving `dev/test.sh`, but I didn't want to make what seemed like intrusive changes without asking first. I figured adding a new test runner on top of `dev/test.sh` would be less intrusive. What niceties would you, @saulpw, and other maintainers support adding to `dev/test.sh`? These changes stand out to me as options:

1. Running each test regardless of if other tests fail.
2. Requiring a clean repo to avoid false results. I can imagine times when this is not desirable, though for me, making a temporary work-in-progress commit is a reasonable workaround.
3. Cleaner output (colors, failure summary at bottom, sending test output to log files)
4. Support for multiple test files passed as command-line arguments.

**Comment by tsibley** [comment]
> Please remove the commit with the added goldens regardless.

Re-pushed with that commit removed.

**Comment by anjakefala** [comment]
Thanks so much @tsibley! That makes a lot more sense now.

I did not get a chance to look too closely yet, because *mornings*. =) Serves me right. I do agree that a script like that sounds useful! I will review it for inclusion.

I have *one* request before I start reviewing! Can you please rename it run `dev/run-tests-individually`? This is so that we only need a single TAB autocomplete to run either of them.

**Comment by saulpw** [comment]
Hey @tsibley, this is really helpful!  I defer to @anjakefala on testing in general.  About your suggestions, for me these sound good:

- Running each test regardless of if other tests fail.
- Cleaner output (colors, failure summary at bottom, sending test output to log files)
- Support for multiple test files passed as command-line arguments.

This one, though, would get in the way of me fixing tests so I'd prefer it not be added:

- Requiring a clean repo to avoid false results.

**Comment by tsibley** [comment]
@saulpw :+1: I'll incorporate those into `./dev/test.sh` directly then in the next day or two.

@anjakefala :city_sunrise: :sleeping: indeed :-)

> Can you please rename it run dev/run-tests-individually? This is so that we only need a single TAB autocomplete to run either of them.

Done! The existing `./dev/testdiff.sh` will still make `./dev/test<TAB>` ambiguous, though.

**Comment by anjakefala** [comment]
@tsibley ahhh, thanks for letting me know!

**Comment by anjakefala** [comment]
> I'll incorporate those into ./dev/test.sh directly then in the next day or two.

When this is done, and I give it a once-over, this will be very joyfully merged in. =)

Thank you, Tom. :blush:

**Comment by anjakefala** [comment]
Hi @tsibley! Would you be okay with us merging this as-is?

If you ever budget time aside to incorporate it into `dev/test.sh`, you can incorporate it in a fresh PR. =) But this is great already.

**Comment by tsibley** [comment]
@anjakefala Yes, that's fine with me! Sorry I haven't had the time I thought I did to rework it. :-/ I started in on it a day or two after saying I would (nearly 3 weeks ago now :scream:), but then ran into some issues I didn't have to consider in the separate script, and then other things pulled me away.

It sometimes takes a long while, but I usually come back to my wips eventually. :-) If I do, and it's still relevant, I'll re-submit it then.

**Comment by anjakefala** [comment]
@tsibley Absolutely *no* worries, I completely get it! In the meanwhile, we can get this already-awesome PR in. =D

**Pull request merged**
